Implement the IRobotControl interface

GiulioRomualdi opened the pull request:
This PR introduces the `IRobotControl` interface. The interface can be used to control a robot. The following control modes are available:
1. Position
2. PositionDirect
3. Velocity
4. Torque

The user can set the desired joint values by using two functions
https://github.com/dic-iit/bipedal-locomotion-framework/blob/a791bc9c9a5efb3a97e0c8e2a7989b7a3de977cb/src/RobotInterface/include/BipedalLocomotion/RobotInterface/IRobotControl.h#L66-L91
The former can be used to set each joint with with a different control mode, in latter the all the joints will be controlled with the same control mode. 

Finally, a Yarp Implementation is provided

traversaro commented on src/RobotInterface/YarpImplementation/src/YarpImplementation.cpp:
```diff
@@ -0,0 +1,554 @@
+            }
+
+            // Yarp wants the quantities in degrees
+            double scaling = 180 / M_PI;
```
Nitpick: YARP support also prismatic joints, that instead do not want any conversion. We can ignore those, but I suggest at least to leave a comment on this.

traversaro replied:
Hi @GiulioRomualdi , if the conversation is marked as solved but I am not sure why.

traversaro replied:
Ah I see, the comment in https://github.com/dic-iit/bipedal-locomotion-framework/pull/97/files#diff-4a8fdd0d7ae7ceafaa3d549e651d9d0fR29 .

traversaro commented:
Just to understand, the `IRobotControl` implementation for YARP has the name `BipedalLocomotion::RobotInterface::YarpImplementation`, but how is the YARP sensor bridge implementation be called?

prashanthr05 commented on src/RobotInterface/YarpImplementation/src/YarpImplementation.cpp:
```diff
@@ -0,0 +1,554 @@
+    std::pair<int, double> getWorstPositionDirectError(Eigen::Ref<const Eigen::VectorXd> desiredJointValues,
+                                                       Eigen::Ref<const Eigen::VectorXd> jointPositions) const
+    {
+        // clear the std::pair
+        std::pair<int, double> worstError{0, 0.0};
```
I suppose the pair describes the joint index and the error? Could you make the comment more descriptive?

prashanthr05 commented on src/RobotInterface/YarpImplementation/src/YarpImplementation.cpp:
```diff
@@ -0,0 +1,554 @@
+                                                     - this->positionFeedback[indeces[i]]);
+
+                    constexpr double scaling = 180 / M_PI;
+                    refSpeeds[i] = std::max(3.0, scaling * (jointError / this->positioningDuration));
```
What's the magic number 3.0?

prashanthr05 commented on src/RobotInterface/YarpImplementation/src/YarpImplementation.cpp:
```diff
@@ -0,0 +1,554 @@
+        m_pimpl->controlModes = controlModes;
+        if (!m_pimpl->setControlModes(m_pimpl->controlModes))
+        {
+            std::cerr << "[YarpImplementation::setReferences] Unable to switch in "
+                         "position-direct control mode."
```
```suggestion
            std::cerr << "[YarpImplementation::setReferences] Unable to switch in "
                         "desired control mode."
```

GiulioRomualdi commented on src/RobotInterface/YarpImplementation/src/YarpImplementation.cpp:
```diff
@@ -0,0 +1,554 @@
+    yarp::dev::IVelocityControl* velocityInterface{nullptr}; /**< Velocity control interface. */
+    yarp::dev::ITorqueControl* torqueInterface{nullptr}; /**< Torque control interface. */
+    yarp::dev::IControlMode* controlModeInterface{nullptr}; /**< Control mode interface. */
+    yarp::dev::IAxisInfo* axisInfoInterface{nullptr}; /**< Axis info interface. */
```
If in the future we need it I can implement it :smile:

GiulioRomualdi commented:
> Just to understand, the IRobotControl implementation for YARP has the name BipedalLocomotion::RobotInterface::YarpImplementation, but how is the YARP sensor bridge implementation be called?

I renamed the class into `RobotInterfaceYarpImplementation`. 

I've also addressed @prashanthr05 and @traversaro  comments

prashanthr05 commented:
> > Just to understand, the IRobotControl implementation for YARP has the name BipedalLocomotion::RobotInterface::YarpImplementation, but how is the YARP sensor bridge implementation be called?
> 
> I renamed the class into `RobotInterfaceYarpImplementation`.
> 
> I've also addressed @prashanthr05 and @traversaro comments

It was renamed to `YarpRobotControl` :+1:
